### PR TITLE
API: Change matplotlib formatter registration

### LINKED ISSUE #18720: VIS: don't modify matplotlib units registry globally
Possible follow-up for https://github.com/pandas-dev/pandas/pull/18307, see comment there: https://github.com/pandas-dev/pandas/pull/18307#issuecomment-347911487 and https://github.com/pandas-dev/pandas/pull/18307#issuecomment-345637950 for the details.

Idea would be to only register our converters when we are plotting ourselves, so people plotting directly with matplotlib (`plt.plot(..)` or `ax.plot(..)` don't use our converters by default (they can still register them manually with the option).

### PR DESCRIPTION
**Pull request by TomAugspurger**
This PR does two things

1. Enforces the deprecation of pandas auto registering our formatters.
2. Provides a new option, "auto", which will only apply our formatters to our plots. By default, the value will be "auto".

Background: pandas used to register formatters for Period, datetime, date, and time. This formatters affected plots created using matplotlib only, not pandas. For various reasons (including import time) we don't want to do this on an `import pandas`.

A minimal change would be to just not do the registration on `import pandas`. Users would either manually do it, or pandas does it automatically when we do DataFrame.plot. The downside of pandas doing it automatically is then in the following

```
fig1 = plt.plot(x, y)
DataFrame.plot(...)  # registers formatters
fig2 = plt.plot(x, y)
```

fig1 and fig2 will be different because pandas has implicilty registered formatters.

The `auto` option will automatically register the formatters, make the plot, then deregister the formatters.

Closes https://github.com/pandas-dev/pandas/issues/18720

### CODE REVIEW

File: pandas/tests/plotting/test_converter.py [hunk 1]
```diff
@@ -71,27 +59,12 @@ def test_register_by_default(self):
         call = [sys.executable, "-c", code]
         assert subprocess.check_call(call) == 0
 
-    def test_warns(self):
-        plt = pytest.importorskip("matplotlib.pyplot")
```
Comment (jreback) [inline pandas/tests/plotting/test_converter.py]: is there a way to check that we *have not* registered things?

Comment (TomAugspurger) [reply pandas/tests/plotting/test_converter.py]: Hmm, we do that indirectly with the code check... If we haven't imported matplotlib, then we know we haven't registered anything.

File: pandas/plotting/_matplotlib/converter.py [hunk 2]
```diff
@@ -55,13 +56,43 @@ def get_pairs():
     return pairs
 
 
-def register(explicit=True):
-    # Renamed in pandas.plotting.__init__
-    global _WARN
+def pandas_converters_deco(func):
```
Comment (WillAyd) [inline pandas/plotting/_matplotlib/converter.py]: Is there a reason why this needs to be a decorator? Perhaps a style preference but I would find it easier to grok if the context manager was used where required in the actual plotting methods

File: pandas/plotting/_matplotlib/converter.py [hunk 3]
```diff
@@ -55,13 +56,43 @@ def get_pairs():
...
-        _WARN = False
+    return wrapper
 
+
+@contextlib.contextmanager
+def pandas_converters():
+    """
+    Context manager registering pandas' converters for a plot.
+
+    When
```
Comment (WillAyd) [inline pandas/plotting/_matplotlib/converter.py]: Typo?

File: pandas/plotting/_matplotlib/converter.py (hunk 2 above)
Comment (TomAugspurger) [reply pandas/plotting/_matplotlib/converter.py]: No need. I mainly did it for a smaller diff / less indentation :)

Happy to go either way here.

**Comment by TomAugspurger** [comment]
All green here now.

**Review by jreback — APPROVED** [review]
lgtm. @datapythonista @jorisvandenbossche

**Review by jorisvandenbossche — APPROVED** [review]
Looks good to me. The auto behaviour seems to work nicely based on a quick test

File: doc/source/user_guide/visualization.rst [hunk 4]
```diff
@@ -1190,6 +1190,18 @@ with "(right)" in the legend. To turn off the automatic marking, use the
...
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+.. versionchanged:: 1.0.0
+
+Pandas provides custom formatters for timeseries plots. These change the
+formatting of the axis labels for dates and times. By default,
+the custom formatters are applied only to plots created by pandas with
+:meth:`DataFrame.plot` or :meth:`Series.plot`. To have them apply to all
+plots, including those made by matplotlib, set the option
+``pd.options.matplotlib.register_converters = True``.
```
Comment (jorisvandenbossche) [inline doc/source/user_guide/visualization.rst]:
```suggestion
``pd.options.plotting.matplotlib.register_converters = True``.
```

Also mention the `pd.plotting.register_matplotlib_converters()` alternative?

File: doc/source/whatsnew/v1.0.0.rst [hunk 5]
```diff
@@ -174,6 +174,10 @@ Backwards incompatible API changes
 
    pd.arrays.IntervalArray.from_tuples([(0, 1), (2, 3)])
 
+- Changed the default configuration value for ``options.matplotlib.register_converters`` from ``True`` to ``"auto"`` (:issue:`18720`).
+  Now, pandas custom formatters will only be applied to plots created by pandas, through :meth:`~DataFrame.plot`.
+  Previously, pandas' formatters would be applied to all plots created *after* a :meth:`~DataFrame.plot`.
```
Comment (jorisvandenbossche) [inline doc/source/whatsnew/v1.0.0.rst]: Add a reference to the longer section below?

**Review by jorisvandenbossche — COMMENTED** [review]
Small doc comment

**Review by datapythonista — APPROVED** [review]
Great PR, nice work. Lgtm, just an idea on implementation and naming, but happy to see this merged with or without thar.

File: pandas/plotting/_matplotlib/converter.py [hunk 6]
```diff
@@ -55,13 +56,42 @@ def get_pairs():
     return pairs
 
 
-def register(explicit=True):
-    # Renamed in pandas.plotting.__init__
-    global _WARN
+def pandas_converters_deco(func):
```
Comment (datapythonista) [inline pandas/plotting/_matplotlib/converter.py]: I think `register_pandas_matplotlib_converters` could be a more clear name.

Also, if the context manager is not going to be used more than in the decorator, I would probably implement its code in the decorator (or move the context manager inside the decorator)/.

File: doc/source/whatsnew/v1.0.0.rst (hunk 5 above)
Comment (jorisvandenbossche) [reply doc/source/whatsnew/v1.0.0.rst]: I meant the section below in the whatsnew (but a link to the docs is actually also useful). There are now 2 entries for this change, and it is not really clear to me how they relate to each other (the one seems to explain it in more detail)

Comment (TomAugspurger) [reply doc/source/whatsnew/v1.0.0.rst]: Oh, whoops.

**Comment by TomAugspurger** [comment]
Should be good to go.

**Comment by jorisvandenbossche** [comment]
Thanks!

**Pull request merged**